expo/expo
# [web][docs] Add supports for push notification in Expo for Web

# Why

Major web browsers (Chrome and Firefox) now all support standardized web push notifications (https://developer.mozilla.org/en-US/docs/Web/API/Push_API). With this PR, developers can use their existing code to also support push notifications in expo-web.

# How

- Added `Notifications.getExpoPushTokenAsync`, `Notifications.getDevicePushTokenAsync`, and `Notifications.addListener` supports for Expo for Web.
- Added docs "Using VAPID for Web Push Notifications".
- Add `notification.vapidPublicKey` in `app.json` and also supports `notification.icon` in `app.json`.
- Add supports for `icon`, `webPath`, `vapidSubject`, `_tag`, and `_renotify` in individual push messages.

This PR is complemented by https://github.com/expo/expo-cli/pull/844 and https://github.com/expo/universe/pull/3610.

# Test Plan

1. Clone expo-cli branch https://github.com/expo/expo-cli/pull/844 as well as the server branch https://github.com/expo/universe/pull/3610.
2. Upload/generate the VAPID keys and VAPID subject following the instruction in `docs/pages/versions/unversioned/guides/using-vapid.md`. Also add `owner` and `notification.vapidPublicKey` in `app.json`.
3. `<path>/expo-cli/node_modules/.bin/expo-cli start`
4. Open http://localhost:19006/ in the browser. (To use the IP address, follow the instruction here: expo/expo#4515)
5. Call `Notifications.getExpoPushTokenAsync()` in your client app and you should get a Expo token. (`native-component-list` should also work as expected without any code changes too.)
6. Send push notifications as usual - e.g. use https://expo.io/dashboard/notifications.
